File: dictionary/javascript/formatting.js
/*
 * Formatting helpers: numbers, currency, bytes, dates, and text layout
 * for reports and user-facing output.
 */

export function formatNumber(value) {
  return value.toLocaleString('en-US');
}

export function formatCurrency(amount, currency = 'USD', locale = 'en-US') {
  return new Intl.NumberFormat(locale, {
    style: 'currency',
    currency,
  }).format(amount);
}

export function formatPercent(value, decimals = 1) {
  return value.toFixed(decimals) + '%';
}

export function formatBytes(bytes) {
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  let value = bytes;
  let unitIndex = 0;
  while (value >= 1024 && unitIndex < units.length - 1) {
    value /= 1024;
    unitIndex += 1;
  }
  const digits = unitIndex === 0 ? 0 : 1;
  return value.toFixed(digits) + ' ' + units[unitIndex];
}

export function formatMillis(milliseconds) {
  const seconds = Math.floor(milliseconds / 1000);
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  const secs = seconds % 60;
  return String(hours).padStart(2, '0') + ':' +
    String(minutes).padStart(2, '0') + ':' +
    String(secs).padStart(2, '0');
}

export function pluralize(count, singular, plural) {
  if (count === 1) return singular;
  return plural || singular + 's';
}

export function ordinal(number) {
  const remainder = number % 100;
  if (remainder >= 11 && remainder <= 13) return number + 'th';
  const suffix = { 1: 'st', 2: 'nd', 3: 'rd' }[number % 10] || 'th';
  return number + suffix;
}

export function formatPhone(digits) {
  const cleaned = digits.replace(/\D/g, '');
  if (cleaned.length === 10) {
    return '(' + cleaned.slice(0, 3) + ') ' + cleaned.slice(3, 6) + '-' + cleaned.slice(6);
  }
  return cleaned;
}

export function formatDateShort(date) {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  return year + '-' + month + '-' + day;
}

export function formatDateTime(date) {
  return formatDateShort(date) + ' ' +
    String(date.getHours()).padStart(2, '0') + ':' +
    String(date.getMinutes()).padStart(2, '0');
}

export function truncateMiddle(text, maxLength) {
  if (text.length <= maxLength) return text;
  const keep = Math.floor((maxLength - 3) / 2);
  return text.slice(0, keep) + '...' + text.slice(-keep);
}

export function formatList(items, conjunction = 'and') {
  if (items.length === 0) return '';
  if (items.length === 1) return String(items[0]);
  if (items.length === 2) return items[0] + ' ' + conjunction + ' ' + items[1];
  return items.slice(0, -1).join(', ') + ', ' + conjunction + ' ' + items[items.length - 1];
}

export function indent(text, spaces = 4) {
  const prefix = ' '.repeat(spaces);
  return text.split('\n').map((line) => prefix + line).join('\n');
}

export function formatLargeNumber(value) {
  if (Math.abs(value) < 1000) return String(value);
  const units = ['K', 'M', 'B', 'T'];
  let scaled = value;
  let unitIndex = -1;
  while (Math.abs(scaled) >= 1000 && unitIndex < units.length - 1) {
    scaled /= 1000;
    unitIndex += 1;
  }
  return scaled.toFixed(1).replace(/\.0$/, '') + units[unitIndex];
}

export function percentChange(before, after) {
  if (before === 0) return 'n/a';
  const change = ((after - before) / Math.abs(before)) * 100;
  return (change >= 0 ? '+' : '') + change.toFixed(1) + '%';
}

export function titleCase(text) {
  return text
    .split(/\s+/)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join(' ');
}

export function formatRating(value, max = 5) {
  const clamped = Math.max(0, Math.min(value, max));
  const full = '★'.repeat(Math.round(clamped));
  const empty = '☆'.repeat(max - Math.round(clamped));
  return full + empty + ' ' + clamped.toFixed(1);
}

export function formatThroughput(bytes, seconds) {
  if (seconds <= 0) return '0 B/s';
  return formatBytes(bytes / seconds) + '/s';
}

export function formatHex(value, width = 2) {
  return value.toString(16).toUpperCase().padStart(width, '0');
}

export function formatCoordinates(latitude, longitude) {
  const latDirection = latitude >= 0 ? 'N' : 'S';
  const lonDirection = longitude >= 0 ? 'E' : 'W';
  return Math.abs(latitude).toFixed(4) + '°' + latDirection + ', ' +
    Math.abs(longitude).toFixed(4) + '°' + lonDirection;
}

export function formatDurationShort(seconds) {
  if (seconds < 60) return seconds + 's';
  if (seconds < 3600) return Math.round(seconds / 60) + 'm';
  return Math.round(seconds / 3600) + 'h';
}

export function formatPercentile(label, value) {
  return label + ': ' + value.toFixed(1) + 'ms';
}

export function formatScore(value, max) {
  const percentage = max > 0 ? (value / max) * 100 : 0;
  return value + '/' + max + ' (' + percentage.toFixed(0) + '%)';
}

File: dictionary/lua/formatting.lua
-- pad_number zero-pads an integer to a fixed width.
local function pad_number(value, width)
	local text = tostring(math.abs(value))
	if value < 0 then
		return "-" .. string.rep("0", math.max(0, width - #text)) .. text
	end
	return string.rep("0", math.max(0, width - #text)) .. text
end

-- comma_separate inserts thousands separators into an integer.
local function comma_separate(value)
	local negative = value < 0
	local digits = tostring(math.abs(value))
	local out = {}
	local count = 0
	for i = #digits, 1, -1 do
		count = count + 1
		out[#out + 1] = digits:sub(i, i)
		if count % 3 == 0 and i > 1 then
			out[#out + 1] = ","
		end
	end
	local joined = table.concat(out):reverse()
	if negative then
		return "-" .. joined
	end
	return joined
end

-- format_percent renders a ratio as a percentage with one decimal.
local function format_percent(ratio)
	return string.format("%.1f%%", ratio * 100)
end

-- format_bytes renders a byte count in human-readable units.
local function format_bytes(bytes)
	local units = { "B", "KB", "MB", "GB", "TB" }
	local value = bytes
	local unit = 1
	while value >= 1024 and unit < #units do
		value = value / 1024
		unit = unit + 1
	end
	if unit == 1 then
		return string.format("%d B", bytes)
	end
	return string.format("%.1f %s", value, units[unit])
end

-- format_duration renders milliseconds as clock-style text.
local function format_duration(ms)
	local total = math.floor(ms / 1000)
	local hours = math.floor(total / 3600)
	local minutes = math.floor((total % 3600) / 60)
	local seconds = total % 60
	if hours > 0 then
		return string.format("%dh %02dm", hours, minutes)
	end
	if minutes > 0 then
		return string.format("%dm %02ds", minutes, seconds)
	end
	return string.format("%ds", seconds)
end

-- align_right pads text on the left to a column width.
local function align_right(text, width)
	if #text >= width then
		return text
	end
	return string.rep(" ", width - #text) .. text
end

-- align_left pads text on the right to a column width.
local function align_left(text, width)
	if #text >= width then
		return text
	end
	return text .. string.rep(" ", width - #text)
end

-- pluralize appends an "s" unless the count is one.
local function pluralize(count, singular)
	if count == 1 then
		return string.format("%d %s", count, singular)
	end
	return string.format("%d %s", count, singular .. "s")
end

-- format_money renders cents as a currency string.
local function format_money(cents)
	local negative = cents < 0
	local absolute = math.abs(cents)
	local dollars = math.floor(absolute / 100)
	local remainder = absolute % 100
	local sign = negative and "-" or ""
	return string.format("%s$%d.%02d", sign, dollars, remainder)
end

-- truncate_middle keeps the head and tail of a long string.
local function truncate_middle(s, max_len)
	if #s <= max_len then
		return s
	end
	local keep = math.floor((max_len - 1) / 2)
	return s:sub(1, keep) .. "..." .. s:sub(-keep)
end

-- wrap_brackets surrounds a value with a configurable pair.
local function wrap_brackets(value, open, close)
	return open .. value .. close
end

-- prefix_lines prepends a marker to each line of text.
local function prefix_lines(text, marker)
	local result = {}
	for line in text:gmatch("[^\n]*") do
		result[#result + 1] = marker .. line
	end
	return table.concat(result, "\n")
end

-- format_key_value renders "key=value" joined by a separator.
local function format_key_value(key, value, sep)
	return key .. (sep or "=") .. value
end

-- escape_html escapes the five HTML-significant characters.
local function escape_html(text)
	local entities = { ["&"] = "&amp;", ["<"] = "&lt;", [">"] = "&gt;", ['"'] = "&quot;", ["'"] = "&#39;" }
	return (text:gsub("[&<>\"']", entities))
end

-- escape_shell_arg quotes an argument for POSIX shells.
local function escape_shell_arg(arg)
	return "'" .. arg:gsub("'", "'\\''") .. "'"
end

-- indent_block indents every line of text by a number of spaces.
local function indent_block(text, indent)
	return prefix_lines(text, string.rep(" ", indent))
end

-- join_oxford joins items with commas and a final "and".
local function join_oxford(items)
	if #items == 0 then
		return ""
	end
	if #items == 1 then
		return items[1]
	end
	if #items == 2 then
		return items[1] .. " and " .. items[2]
	end
	local head = table.concat(items, ", ", 1, #items - 1)
	return head .. ", and " .. items[#items]
end

-- pad_center centres text within a width using spaces.
local function pad_center(text, width)
	if #text >= width then
		return text
	end
	local left = math.floor((width - #text) / 2)
	local right = width - #text - left
	return string.rep(" ", left) .. text .. string.rep(" ", right)
end

-- format_table aligns rows of cells into fixed-width columns.
local function format_table(rows)
	if #rows == 0 then
		return {}
	end
	local widths = {}
	for _, row in ipairs(rows) do
		for i, cell in ipairs(row) do
			widths[i] = math.max(widths[i] or 0, #tostring(cell))
		end
	end
	local result = {}
	for _, row in ipairs(rows) do
		local cells = {}
		for i, cell in ipairs(row) do
			cells[i] = align_left(tostring(cell), widths[i])
		end
		result[#result + 1] = table.concat(cells, "  ")
	end
	return result
end

-- format_number aligns a column of numbers to the right.
local function format_number(numbers)
	local width = 0
	for _, n in ipairs(numbers) do
		width = math.max(width, #tostring(n))
	end
	local result = {}
	for i, n in ipairs(numbers) do
		result[i] = align_right(tostring(n), width)
	end
	return result
end

-- numbered_list renders items as "1. item" lines.
local function numbered_list(items)
	local result = {}
	for i, item in ipairs(items) do
		result[#result + 1] = string.format("%d. %s", i, item)
	end
	return table.concat(result, "\n")
end
